src/main/java/spoon/reflect/visitor/ImportCleaner.java

```diff
@@ -201,6 +204,12 @@ void addImport(CtReference ref) {
 			}
 		}
 
+		private boolean isPackageImportedViaWildcardAndUnresolved(CtPackageReference packageReference) {
+			return compilationUnit.getImports().stream().anyMatch(
+					ctImport -> ctImport.toString().contains(packageReference.toString())
+							&& ctImport.getImportKind() == CtImportKind.UNRESOLVED);
```

Comment:
When `toString()` is used for matching like this, my spidey sense always tingles a bit.

Imagine that `packageReference` points to `java.util`, and `ctImport` is `java.util.concurrent.*`. Don't we get a false positive there? I.e. `"java.util.concurrent.*".contains("java.util")` is true, and `java.util.concurrent.*` is a wildcard import.

Reply:
I delve into `ImportCleaner` further and I was wondering about its existence. We use `ImportCleaner` to add/remove `import` statements, but I can't figure out why that is a function of Spoon. If the imports in a compilation unit are missing or extra, that will cause a compilation error and spoon is not expected to behave correctly on such models.

You are right that this case will be a false-positive. The `ImportCleaner` will not add import, say `java.util.List`, if `java.util.concurrent.*` is present and `List` interface is used somewhere in the code. But is it Spoon's job to do so? The program would not compile even before it's processed with Spoon.

Anyway, of course, we need a better way to know if the wildcard import exists corresponding to the `packageReference`. It might be sufficient to look at the top-level type and then compare but since the imports are unresolved, we will have to do that via splitting string and that sounds ugly to me.

Reply:
> You are right that this case will be a false-positive. The ImportCleaner will not add import, say java.util.List, if java.util.concurrent.* is present and List interface is used somewhere in the code. But is it Spoon's job to do so? The program would not compile even before it's processed with Spoon.

You're thinking of this from the perspective of this being the _input_, but Spoon is made for AST transformations. The purpose of auto-import functionality is to automatically add imports after one has transformed the AST. For example, in a Spoon model, I can insert a variable declaration with type `java.util.List`, but I don't want to find this in the printed output:

```java
java.util.List<Integer> myNewList = new java.util.ArrayList<>();
```

No one writes Java like that. Rather, I want this:

```java
List<Integer> myNewList = new ArrayList<>();
```

Since imports aren't _really_ part of the AST (they belong to the compilation unit), it's kind of awkward to get that right.

As a concrete example closer to what you work with, look at the `XxeProcessingProcessor` in Sorald (rule 2755). It inserts references to the `java.xml.XMLConstants` class, which must sometimes be imported (see for example [this PR](https://github.com/redhat-developer/rsp-server/pull/619/files)). But if you look at the implementation of that processor, it does nothing with import statements, that's handled later during output processing.

Reply:
That makes sense.

> As a concrete example closer to what you work with, look at the XxeProcessingProcessor in Sorald (rule 2755). It inserts references to the java.xml.XMLConstants class, which must sometimes be imported (see for example this PR). But if you look at the implementation of that processor, it does nothing with import statements, that's handled later during output processing.

Yes, I disabled the `ImportCleaner` and saw a flurry of failed test cases. Rules like 2116 also required import statements (`java.util.Arrays`) to be added.